src/libostree/ostree-core.c

```diff
@@ -302,6 +303,47 @@ ostree_validate_collection_id (const char *collection_id, GError **error)
   return TRUE;
 }
 
+static int
+compare_xattrs (const void *a_pp, const void *b_pp)
+{
+  GVariant *a = *(GVariant **)a_pp;
+  GVariant *b = *(GVariant **)b_pp;
+  const char *name_a;
+  const char *name_b;
+  g_variant_get (a, "(^&ay@ay)", &name_a, NULL);
+  g_variant_get (b, "(^&ay@ay)", &name_b, NULL);
+
+  return strcmp (name_a, name_b);
+}
+
+// Sort xattrs by name
+static GVariant *
+canonicalize_xattrs (GVariant *xattrs)
```

Comment:
Yikes.  This function is complicated, but I honestly can't think of any other way to do it more simply...

One thing I'd consider is scanning the "input" to see if it's already in order (trivially true also for the common case of only one item) and return a ref directly in that case.

Reply:
> Yikes. This function is complicated, but I honestly can't think of any other way to do it more simply...

Yeah...this is where GVariant in C specifically is really painful. At one point Alex was working on a "gvariant compiler" IIRC that generated accessors. In Rust there's https://docs.rs/gvariant/latest/gvariant/ but none of this code is going to be in Rust soon sadly.